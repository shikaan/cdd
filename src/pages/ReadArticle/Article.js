import React, {Component} from "react";
import styled from "styled-components";
import "prismjs/themes/prism-dark.css";

const paragraphFontSize = (theme) => theme.typography.baseFontSize.multiply(1.125);
const h1FontSize = (theme) => theme.typography.baseFontSize.multiply(2);
const h2FontSize = (theme) => theme.typography.baseFontSize.multiply(1.75);
const h3FontSize = (theme) => theme.typography.baseFontSize.multiply(1.5);

const ArticleBody = styled.article(({theme}) => `
  padding: ${theme.templateVariables.horizontalPadding}
  
  p {
    font-family: ${theme.typography.secondaryFont.fontFamily};
    line-height: ${paragraphFontSize(theme).multiply(1.625)};
    font-weight: 400;
    letter-spacing: ${paragraphFontSize(theme).multiply(0.01)};
    font-size: ${paragraphFontSize(theme)};
    padding: ${paragraphFontSize(theme).multiply(0.75)} 0;
    color: ${theme.color.dark600};
    
    code {
      white-space: nowrap;
    }
  }
  
  h1 {
    font-family: ${theme.typography.primaryFont.fontFamily};
    line-height: ${h1FontSize(theme).multiply(1.5)};
    font-weight: 700;
    font-size: ${h1FontSize(theme)};
    color: ${theme.color.dark600};
  }
  
  h2 {
    font-family: ${theme.typography.primaryFont.fontFamily};
    line-height: ${h2FontSize(theme).multiply(1.5)};
    font-weight: 700;
    font-size: ${h2FontSize(theme)};
    color: ${theme.color.dark600} 
  }
  
  h3, h4, h5 {
    font-family: ${theme.typography.primaryFont.fontFamily};
    line-height: ${h3FontSize(theme).multiply(1.5)};
    font-size: ${h3FontSize(theme)};
    font-weight: 400;
    color: ${theme.color.dark600} 
  }
  
  sup {
    vertical-align: text-bottom;
    font-size: ${paragraphFontSize(theme).multiply(.75)};
    padding: ${paragraphFontSize(theme).multiply(.25)};
  }
  
  em {
    font-style: italic;
  }
  
  strong {
    font-weight: 700;
  }
  
  img {
    margin: auto;
  }
  
  ul {
    list-style-type: disc;
    list-style-position: inside;
    padding: ${paragraphFontSize(theme)};
    
    ul {
      list-style-type: circle;
      list-style-position: inside;
      padding: 0 ${paragraphFontSize(theme)};
    }
    
    li {
      padding: ${paragraphFontSize(theme).multiply(.25)} 0;
      > p {
        display: inline;
      }
    }
  }
  
  blockquote {
    border-left: 4px solid ${theme.color.brand400};
    padding: 0 ${theme.typography.baseFontSize};
    margin: ${theme.typography.baseFontSize} 0;
    background: ${theme.color.brand100};
  }
  
  .footnotes {
    ol {
      list-style-type: decimal;
      list-style-position: inside;
      
      li {
        padding: ${paragraphFontSize(theme).multiply(.5)} 0;
        
        p { 
          font-size: ${paragraphFontSize(theme)};
          display: inline;
        }
      }
    }
  }
`);

class Article extends Component {
  render() {
    const {post} = this.props;

    return (
      <ArticleBody dangerouslySetInnerHTML={{__html: post.html}}/>
    );
  }
}

export default Article;
