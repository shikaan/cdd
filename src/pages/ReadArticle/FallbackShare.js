import React from 'react';
import styled from "styled-components";

import {FacebookShareButton, FacebookIcon, TwitterShareButton, TwitterIcon} from "react-share";

import Divider from '~components/Divider'
import Icon from '~components/Icon'

const Overlay = styled.div(({theme, visible}) => `
  top: 0;
  left: 0;
  bottom: 0;
  right: 0;
  position: fixed;
  display: ${visible ? 'flex' : 'none'};
  height: 100%;
  width: 100%;
  background-color: ${theme.color.grey};
  z-index: ${theme.zIndexRank.top};
  cursor: pointer;
`)

const Menu = styled.nav(({theme}) => `
  align-self: flex-end;
  background: ${theme.color.paleGrey};
  display: flex;
  flex-direction: column;
  width: 100vw;
`)

const MenuTitle = styled.span(({theme}) => `
  flex: 1;
  font-size: ${theme.typography.baseFontSize.multiply(1)};
  color: ${theme.color.black};
  padding: ${theme.typography.baseFontSize};
`)

const MenuItems = styled.ul`
  flex: 1
`

const MenuItem = styled.li(({theme}) => `
  list-style:none
  padding: ${theme.typography.baseFontSize};
  
  & > * {
    display: flex;
  }
`)

const MenuItemLabel = styled.span(({theme, size}) => `
  line-height: ${size}px;
  padding: 0 ${theme.typography.baseFontSize};
  flex: 1;
`)

const CloseShareButton = styled.div``

const CloseShareIcon = styled(Icon)(({size}) => `
  line-height: ${size}px;
  text-align: center;
  width: ${size}px;
`)

const FallbackShare = (props) => {
  const {onClose, visible} = props

  return (
    <Overlay onClick={onClose} visible={visible}>
      <Menu>
        <MenuTitle>Share</MenuTitle>
        <Divider/>
        <MenuItems>
          <MenuItem>
            <FacebookShareButton url={window.location.href}>
              <FacebookIcon round size={32}/>
              <MenuItemLabel size={32}>Facebook</MenuItemLabel>
            </FacebookShareButton>
          </MenuItem>
          <MenuItem>
            <TwitterShareButton url={window.location.href}>
              <TwitterIcon round size={32}/>
              <MenuItemLabel size={32}>Twitter</MenuItemLabel>
            </TwitterShareButton>
          </MenuItem>
        </MenuItems>
        <Divider/>
        <MenuItem>
          <CloseShareButton>
            <CloseShareIcon size={32}>
              close
            </CloseShareIcon>
            <MenuItemLabel size={32}>Close</MenuItemLabel>
          </CloseShareButton>
        </MenuItem>
      </Menu>
    </Overlay>
  );
}

export default FallbackShare;
