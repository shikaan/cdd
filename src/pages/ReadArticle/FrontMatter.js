import React, {Component, Fragment} from "react";
import styled from "styled-components";

import Heading from "~components/Heading";
import Image from "~components/Image";
import Overline from "~components/Overline";
import Tag from "~components/Tag";

const Header = styled.header(({theme}) => `
  padding: ${theme.templateVariables.verticalPadding} ${theme.templateVariables.horizontalPadding};
`);

const FrontMatterHeading = styled(Heading)(({theme}) => `
  padding-bottom: ${theme.templateVariables.verticalPadding};
`);

class FrontMatter extends Component {
  render() {
    const {content, post, tags} = this.props;

    return (
      <Fragment>
        <Image fluid={post.coverImage.fluid} alt={post.title}/>
        <Header>
          <Overline>
            {post.createdAt} â€“ {post.timeToRead} {content.shared.readingTime}
          </Overline>

          <FrontMatterHeading level={1} sub={post.description}>
            {post.title}
          </FrontMatterHeading>

          {tags.map(i => <Tag key={i} to={`/search?query=${i}`}>{i}</Tag>)}
        </Header>
      </Fragment>
    );
  }
}

export default FrontMatter;
