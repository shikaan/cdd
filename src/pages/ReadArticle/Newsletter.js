import React, { Component } from "react";
import styled from "styled-components";

import { Size } from "~theme";
import { NEWSLETTER_WEBHOOK_ADDRESS } from "~/constants";

import Button from "~components/Button";
import Heading from "~components/Heading";
import Form from "~components/Form";
import Input from "~components/Input";
import Toaster from "~components/Toaster";

const Section = styled.section(({ theme }) => `
  text-align: center;
  padding: ${theme.templateVariables.horizontalPadding}
`);

const NewsletterHeading = styled(Heading)(({ theme }) => `
  padding-bottom: ${theme.templateVariables.verticalPadding};
`);

const NewsletterForm = styled(Form)(() => `
  min-width: ${new Size(40)};
  max-width: 62%;
  margin: auto;
`);

const InputWrapper = styled.div(({ theme }) => `
  padding: ${theme.templateVariables.horizontalPadding};
`);

class Newsletter extends Component {
  static TOASTER_DURATION = 3000;
  static INITIAL_STATE = {
    email: "",
    showSuccessToaster: false,
    showErrorToaster: false
  };

  inputRef = React.createRef();

  state = { ...Newsletter.INITIAL_STATE };

  saveEmailAddress = (address) => {
    return fetch(NEWSLETTER_WEBHOOK_ADDRESS, { method: "POST", body: JSON.stringify({ email: address }) });
  };

  handleEmailChange = ({ target }) => {
    this.setState({ email: target.value });
  };

  handleSubscribe = async (e) => {
    e.preventDefault();

    const isEmailValid = this.inputRef.current.reportValidity();

    if (isEmailValid) {
      try {
        await this.saveEmailAddress(this.state.email);
        this.handleSubscribeSuccess();
      } catch (e) {
        this.handleSubscribeError();
      }
    }
  };

  handleSubscribeSuccess = () => {
    this.setState({ showSuccessToaster: true }, () => {
      setTimeout(() => {
        this.setState(Newsletter.INITIAL_STATE);
      }, Newsletter.TOASTER_DURATION);
    });
  };

  handleSubscribeError = () => {
    this.setState({ showErrorToaster: true }, (state) => {
      setTimeout(() => {
        const email = state.email;
        this.setState({ ...Newsletter.INITIAL_STATE, email });
      }, Newsletter.TOASTER_DURATION);
    });
  };

  render () {
    const { content } = this.props;
    const { email, showSuccessToaster, showErrorToaster } = this.state;

    return (
      <Section>
        <Toaster visible={showSuccessToaster}>
          {content.newsletter.successMessage}
        </Toaster>
        <Toaster visible={showErrorToaster}>
          {content.newsletter.failMessage}
        </Toaster>
        <NewsletterHeading level={2} sub={content.newsletter.subTitle}>
          {content.newsletter.title}
        </NewsletterHeading>
        <NewsletterForm onSubmit={this.handleSubscribe}>
          <InputWrapper>
            <Input
              ref={this.inputRef}
              type="email"
              required
              placeholder={content.newsletter.inputPlaceholder}
              value={email}
              onChange={this.handleEmailChange}/>
          </InputWrapper>

          <Button context="accent" type="submit">
            {content.newsletter.actionLabel}
          </Button>
        </NewsletterForm>
      </Section>
    );
  }
}

export default Newsletter;
