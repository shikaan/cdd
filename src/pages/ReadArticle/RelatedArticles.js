import React, {Component} from 'react';
import styled from "styled-components";

import {Size} from '~theme'

import Card from '~components/Card'
import Divider from "~components/Divider";
import Heading from "~components/Heading";

const Section = styled.section(({theme}) => `
  padding: ${theme.templateVariables.horizontalPadding}
`)

const UnorderedList = styled.ul(({theme}) => `
  padding: ${theme.templateVariables.verticalPadding} 0;
`)

const ListItem = styled.li(() => `
  min-width: ${new Size(40)};
  max-width: 62%;
  margin: auto;
`)

class RelatedArticles extends Component {
  renderDivider = (index) => {
    const isLastElement = index === this.props.list.length - 1

    return isLastElement ? null : <Divider/>
  }

  render() {
    const {content, list} = this.props

    return (
      <Section>
        <Heading level={2}>
          {content.relatedArticles.title}
        </Heading>
        <UnorderedList>
          {
            list.map(({node}, index) => (
              <ListItem>
                <Card key={index} post={node} content={content}/>
                {this.renderDivider(index)}
              </ListItem>
            ))
          }
        </UnorderedList>
      </Section>
    );
  }
}

export default RelatedArticles;
