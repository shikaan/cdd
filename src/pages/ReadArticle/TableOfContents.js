import React, { useState } from "react";
import styled from "styled-components";
import Icon from "~components/Icon";
import {getMicrocopy} from "~/utils";

const bodyTwoStyle = (theme) => `
  font-family: ${theme.typography.secondaryFont.fontFamily};
  font-weight: 400;
  font-size: ${theme.typography.baseFontSize};
`;

const Content = styled.nav`
  font-family: ${({ theme }) => theme.typography.secondaryFont.fontFamily};
  padding: 0 ${({ theme }) => theme.templateVariables.horizontalPadding};
`;

const ContentHeader = styled.header`
  ${({ theme }) => bodyTwoStyle(theme)}
  font-weight: 700;
  padding: ${({ theme }) =>
    `${theme.templateVariables.verticalPadding} ${theme.templateVariables.horizontalPadding}`};
  cursor: pointer;
`;

const ContentSection = styled.section(
  ({ theme, closed }) => `

  height: ${closed ? 0 : "auto"};
  overflow: hidden;


  padding: 0 ${theme.templateVariables.horizontalPadding} ${
    closed ? 0 : theme.templateVariables.verticalPadding.multiply(2)
  } ${theme.templateVariables.horizontalPadding};

  
  & > ul {
    padding: 0 ${theme.typography.baseFontSize.multiply(.5)};
    list-style-position: inside;
     
    p {
        display: inline;
    }
    
    ul {
      color: ${theme.color.dark500};
    }
    
    a {
      ${bodyTwoStyle(theme)}
      text-decoration: none;
      padding: ${theme.templateVariables.verticalPadding.multiply(.5)} 0;
      display: inline-block;
      width: calc(100% - ${theme.templateVariables.horizontalPadding});
    }
    
    li {      
      font-size: ${theme.typography.baseFontSize};
      list-style-type: disc;
      list-style-position: inside;
      
      li {
        list-style-type: circle;
        padding-left: ${theme.typography.baseFontSize};
      }
     }
  }
`
);

const Caret = styled(Icon)(({ closed }) => `
  float: right;
  ${closed ? "" : "transform: rotate(180deg);"}
`);

const TableOfContents = ({ content, post }) => {
  const [closed, setClosed] = useState(false);

  return (
    <Content>
      <ContentHeader onClick={() => setClosed(!closed)}>
        {content.title}
        <Caret closed={closed} icon="caret" />
      </ContentHeader>
      <ContentSection
        closed={closed}
        dangerouslySetInnerHTML={{
          __html: post.body.childMarkdownRemark.tableOfContents,
        }}
      />
    </Content>
  );
};

export default TableOfContents;
