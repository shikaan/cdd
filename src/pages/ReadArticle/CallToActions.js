import React, { Component } from 'react'
import styled from 'styled-components'

import { REPO_BASE_URL } from '~/constants'

import Icon from '~components/Icon'
import Link from '~components/Link'

import FallbackShare from './FallbackShare'

const Section = styled.section(({ theme }) => `
  padding: 0 ${theme.templateVariables.horizontalPadding.multiply(2)};
  padding-bottom: ${theme.templateVariables.verticalPadding.multiply(2)};
  display: flex;
  justify-content: space-between;
`)

const CTALink = styled(Link)(({ theme }) => `
  color: ${theme.color.coffee};
  font-size: ${theme.typography.baseFontSize.multiply(1.75)}
`)

class CallToActions extends Component {
  state = {
    isFallbackShareVisible: false
  };

  static buildEditUrl (relativeFilePath) {
    return `${REPO_BASE_URL}/edit/master/${relativeFilePath}`
  }

  static useFallbackShare () {
    return !navigator.share
  }

  share = () => {
    const { post } = this.props

    if (CallToActions.useFallbackShare()) {
      this.openFallbackShare()
    } else {
      navigator.share({
        url: window.location.href,
        title: document.title,
        text: post.frontmatter.title
      })
    }
  };

  openFallbackShare = () => {
    this.setState({ isFallbackShareVisible: true })
  };

  closeFallbackShare = () => {
    this.setState({ isFallbackShareVisible: false })
  };

  renderFallbackShare = (shareVisible) => {
    return CallToActions.useFallbackShare()
      ? <FallbackShare onClose={this.closeFallbackShare} visible={shareVisible}/>
      : null
  };

  render () {
    const { content, post } = this.props

    return (
      <Section>
        <CTALink alt={content.callToActions.comment} href={post.frontmatter.commentLink}>
          <Icon>comment</Icon>
        </CTALink>

        <CTALink alt={content.callToActions.edit} href={CallToActions.buildEditUrl(post.fields.relativeFilePath)}>
          <Icon>edit</Icon>
        </CTALink>

        <CTALink alt={content.callToActions.share} onClick={this.share}>
          <Icon>share</Icon>
        </CTALink>

        {this.renderFallbackShare(this.state.isFallbackShareVisible)}
      </Section>
    )
  }
}

export default CallToActions
