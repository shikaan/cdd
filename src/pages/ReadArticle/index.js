import React from "react"
import {graphql} from "gatsby"

import {en as content} from '/static/content/ReadArticle'

import Template from "~templates/Main"

import Divider from "~components/Divider"
import SEO from "~components/seo"

import CallToActions from "./CallToActions";
import FrontMatter from "./FrontMatter";
import TableOfContents from "./TableOfContents";
import Article from "./Article";
import Newsletter from "./Newsletter";

class BlogPostTemplate extends React.Component {
  render() {
    const post = this.props.data.markdownRemark
    const siteTitle = this.props.data.site.siteMetadata.title

    return (
      <Template location={this.props.location} title={siteTitle}>
        <SEO title={post.frontmatter.title} description={post.frontmatter.description}/>

        <FrontMatter post={post} pageContext={this.props.pageContext} content={content}/>
        <Divider/>
        <TableOfContents post={post} content={content}/>
        <Divider/>
        <Article post={post}/>
        <CallToActions post={post} content={content}/>
        <Divider/>
        <Newsletter content={content}/>
      </Template>
    )
  }
}

export default BlogPostTemplate

export const pageQuery = graphql`
  query BlogPostBySlug($path: String!) {
    site {
      siteMetadata {
        title
        author
      }
    }
    markdownRemark(fields: { slug: { eq: $path } }) {
      html
      tableOfContents(
        maxDepth: 2
      )
      fields {
        slug
        relativeFilePath
        readingTime {
          minutes
        }
      }
      frontmatter {
        title
        date(formatString: "MMMM DD, YYYY")
        description
        commentLink
        coverImage {
          childImageSharp {
            fluid {
              base64
              tracedSVG
              aspectRatio
              src
              srcSet
              srcWebp
              srcSetWebp
              sizes
              originalImg
              originalName
              presentationWidth
              presentationHeight
            }
          }
        }
      }
    }
  }
`
