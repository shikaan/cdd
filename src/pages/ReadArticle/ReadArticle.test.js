import React from 'react'
import {shallow} from 'enzyme';
import 'jest-styled-components';

import ReadArticle from ".";

describe('ReadArticle', () => {
  it('renders correctly', () => {
    const wrapper = shallow(<ReadArticle/>)

    expect(wrapper).toMatchSnapshot()
  })
})
