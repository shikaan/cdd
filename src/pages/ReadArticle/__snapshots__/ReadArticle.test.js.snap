// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReadArticle renders correctly 1`] = `
<MainTemplate
  title=""
>
  <SEO
    description=""
    keywords={Array []}
    lang="en"
    meta={Array []}
    title=""
  />
  <FrontMatter
    content={
      Object {
        "callToActions": Object {
          "comment": "Comment on Twitter",
          "edit": "Edit on Github",
          "share": "Share",
        },
        "newsletter": Object {
          "ctaLabel": "Subscribe",
          "failMessage": "Error! Try again in a few moments",
          "inputPlaceholder": "Your email here",
          "subTitle": "Get notified about new articles!",
          "successMessage": "All good!",
          "title": "Enjoying?",
        },
        "relatedArticles": Object {
          "title": "Related Articles",
        },
        "shared": Object {
          "readingTime": "min read",
        },
        "tableOfContents": Object {
          "title": "Content",
        },
      }
    }
  />
  <Styled(styled.hr) />
  <TableOfContents
    content={
      Object {
        "callToActions": Object {
          "comment": "Comment on Twitter",
          "edit": "Edit on Github",
          "share": "Share",
        },
        "newsletter": Object {
          "ctaLabel": "Subscribe",
          "failMessage": "Error! Try again in a few moments",
          "inputPlaceholder": "Your email here",
          "subTitle": "Get notified about new articles!",
          "successMessage": "All good!",
          "title": "Enjoying?",
        },
        "relatedArticles": Object {
          "title": "Related Articles",
        },
        "shared": Object {
          "readingTime": "min read",
        },
        "tableOfContents": Object {
          "title": "Content",
        },
      }
    }
  />
  <Styled(styled.hr) />
  <Article />
  <CallToActions
    content={
      Object {
        "callToActions": Object {
          "comment": "Comment on Twitter",
          "edit": "Edit on Github",
          "share": "Share",
        },
        "newsletter": Object {
          "ctaLabel": "Subscribe",
          "failMessage": "Error! Try again in a few moments",
          "inputPlaceholder": "Your email here",
          "subTitle": "Get notified about new articles!",
          "successMessage": "All good!",
          "title": "Enjoying?",
        },
        "relatedArticles": Object {
          "title": "Related Articles",
        },
        "shared": Object {
          "readingTime": "min read",
        },
        "tableOfContents": Object {
          "title": "Content",
        },
      }
    }
  />
  <Styled(styled.hr) />
  <Newsletter
    content={
      Object {
        "callToActions": Object {
          "comment": "Comment on Twitter",
          "edit": "Edit on Github",
          "share": "Share",
        },
        "newsletter": Object {
          "ctaLabel": "Subscribe",
          "failMessage": "Error! Try again in a few moments",
          "inputPlaceholder": "Your email here",
          "subTitle": "Get notified about new articles!",
          "successMessage": "All good!",
          "title": "Enjoying?",
        },
        "relatedArticles": Object {
          "title": "Related Articles",
        },
        "shared": Object {
          "readingTime": "min read",
        },
        "tableOfContents": Object {
          "title": "Content",
        },
      }
    }
  />
  <Styled(styled.hr) />
  <RelatedArticles
    content={
      Object {
        "callToActions": Object {
          "comment": "Comment on Twitter",
          "edit": "Edit on Github",
          "share": "Share",
        },
        "newsletter": Object {
          "ctaLabel": "Subscribe",
          "failMessage": "Error! Try again in a few moments",
          "inputPlaceholder": "Your email here",
          "subTitle": "Get notified about new articles!",
          "successMessage": "All good!",
          "title": "Enjoying?",
        },
        "relatedArticles": Object {
          "title": "Related Articles",
        },
        "shared": Object {
          "readingTime": "min read",
        },
        "tableOfContents": Object {
          "title": "Content",
        },
      }
    }
    list={Array []}
  />
</MainTemplate>
`;
